mlflow/pipelines/steps/predict.py

```diff
@@ -0,0 +1,163 @@
+                "PROFILE", scored_dataset_profile
+            )
+
+        # Tab #1/2: run summary.
+        (
+            card.add_tab(
+                "Run Summary",
+                """
+                {{ SCORED_DATA_NUM_ROWS }}
+                {{ EXE_DURATION }}
+                {{ LAST_UPDATE_TIME }}
+                """,
+            ).add_markdown(
+                "SCORED_DATA_NUM_ROWS",
+                f"**Number of scored dataset rows:** `{len(scored_df)}`",
+            )
+        )
+
+        return card
+
+    def _run(self, output_directory, _env_manager="conda"):
+        from pyspark.sql.functions import struct
+
+        run_start_time = time.time()
+
+        # Get or create spark session
+        try:
+            spark = _get_active_spark_session()
+        except Exception as e:
+            raise MlflowException(
+                message=(
+                    "Encountered an error while searching for an active Spark session to"
+                    " score dataset with spark UDF. Please create a Spark session and try again."
+                ),
+                error_code=BAD_REQUEST,
+            ) from e
+        if not spark:
+            raise MlflowException(
+                message=(
+                    "No active SparkSession detected to score dataset with spark UDF. "
+                    "Please create a Spark session and try again."
+                ),
+                error_code=BAD_REQUEST,
+            )
+
+        # read cleaned dataset
+        ingested_data_path = get_step_output_path(
+            pipeline_root_path=self.pipeline_root,
+            step_name="ingest_scoring",
+            relative_path=_INPUT_FILE_NAME,
+        )
+        input_sdf = spark.read.parquet(ingested_data_path)
+        if _PREDICTION_COLUMN_NAME in input_sdf.columns:
+            _logger.warning(
+                f"Input scoring dataframe already contains a column '{_PREDICTION_COLUMN_NAME}'. "
+                f"This column will be dropped in favor of the predict output column name."
+            )
+
+        # score dataset
+        model_uri = self.step_config["model_uri"]
+        predict = mlflow.pyfunc.spark_udf(spark, model_uri, env_manager=_env_manager)
+        scored_sdf = input_sdf.withColumn(
+            _PREDICTION_COLUMN_NAME, predict(struct(*input_sdf.columns))
+        )
+
+        # save predictions
+        # note: the current output writing logic allows no overwrites
+        output_format = self.step_config["output_format"]
+        if output_format == "parquet" or output_format == "delta":
+            scored_sdf.coalesce(1).write.format(output_format).save(
+                self.step_config["output_location"]
+            )
+        else:
+            scored_sdf.write.format("delta").saveAsTable(self.step_config["output_location"])
+
+        # predict step artifacts
+        scored_pdf = scored_sdf.toPandas()
+        scored_pdf.to_parquet(
```

Comment:
Let's use Spark to save the full dataset here, instead of saving a downsampled one.